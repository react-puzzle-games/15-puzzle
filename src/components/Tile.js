import React, { Component } from 'react';
import styled from 'styled-components';
import classnames from 'classnames';
import PropTypes from 'prop-types';

class Tile extends Component {
  constructor(props) {
    super(props);

    this.onClick = this.onClick.bind(this);
  }

  onClick() {
    this.props.onClick(this.props);
  }

  render() {
    const { className, number, visible } = this.props;
    const classes = classnames({
      [className]: true,
      'tile-invisible': !visible,
    });

    return (
      <div className={classes} onClick={this.onClick}>
        <span className="tile-number">{number}</span>
      </div>
    );
  }
}

Tile.propTypes = {
  tileId: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  left: PropTypes.number.isRequired,
  top: PropTypes.number.isRequired,
  number: PropTypes.number,
  onClick: PropTypes.func,
  correct: PropTypes.bool,
  visible: PropTypes.bool,
};

Tile.defaultProps = {
  number: 0,
  correct: false,
  visible: true,
};

export default styled(Tile)`
  border: 1px solid #FFD1AA;
  position: absolute;
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  left: ${props => props.left}px;
  top: ${props => props.top}px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: ${props => props.correct ? '#226666' : '#D4726A'};
  transition-property: top, left, background-color;
  transition-duration: .300s;
  transition-timing-function: ease-in;

  .tile-number {
    color: #FFD1AA;
    font-weight: 400;
    font-size: 1.8em;
    user-select: none;
  }

  &.tile-invisible {
    display: none;
  }
`;

export const propTypes = Tile.propTypes;
