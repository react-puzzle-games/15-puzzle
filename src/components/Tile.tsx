import { makeStyles } from "@material-ui/styles";
import React from "react";

const useStyles = makeStyles({
  root: ({ width, height, correct, left, top, visible }: Props) => ({
    display: visible ? "flex" : "none",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    position: "absolute",
    border: "1px solid #FFD1AA",
    width,
    height,
    left,
    top,
    cursor: "pointer",
    backgroundColor: correct ? "#226666" : "#D4726A",
    transitionProperty: "top, left, background-color",
    transitionDuration: ".300s",
    transitionTimingFunction: "ease-in",
  }),

  tileNumber: {
    color: "#FFD1AA",
    fontSize: "1.8em",
    userSelect: "none",
  },
});

type Props = {
  tileId: number;
  number: number;
  onClick: (props: Props) => void;
  width: number;
  height: number;
  correct: boolean;
  left: number;
  top: number;
  visible?: boolean;
};

const Tile = (props: Props) => {
  const { number = 0, onClick } = props;
  const styles = useStyles(props);

  return (
    <div className={styles.root} onClick={() => onClick(props)}>
      <span className={styles.tileNumber}>{number}</span>
    </div>
  );
};

export default Tile;
