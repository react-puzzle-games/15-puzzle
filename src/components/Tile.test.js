import renderer from 'react-test-renderer';
import React from 'react';
import Tile from './Tile';

describe('Tile renders correctly', () => {
  test('given custom number', () => {
    const tree = renderer
      .create(
        <Tile
          tileId={50}
          left={0}
          top={0}
          number={13}
          width={90}
          height={90}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('given default number', () => {
    const tree = renderer
      .create(
        <Tile tileId={18} left={-40} top={-20} width={120} height={120} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('given correct prop', () => {
    const tree = renderer
      .create(
        <Tile
          tileId={13}
          left={0}
          top={50}
          correct={true}
          width={90}
          height={90}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('given visible prop', () => {
    const tree = renderer
      .create(
        <Tile
          tileId={4}
          left={30}
          top={0}
          visible={false}
          width={30}
          height={30}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
