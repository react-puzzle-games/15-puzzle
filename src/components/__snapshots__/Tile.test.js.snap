// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile renders correctly given correct prop 1`] = `
<div
  className="sc-bdVaJa dFyZnU"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    0
  </span>
</div>
`;

exports[`Tile renders correctly given custom number 1`] = `
<div
  className="sc-bdVaJa jodVcS"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    13
  </span>
</div>
`;

exports[`Tile renders correctly given default number 1`] = `
<div
  className="sc-bdVaJa hYSZle"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    0
  </span>
</div>
`;

exports[`Tile renders correctly given visible prop 1`] = `
<div
  className="sc-bdVaJa clEXFS tile-invisible"
  onClick={[Function]}
>
  <span
    className="tile-number"
  >
    0
  </span>
</div>
`;
